x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportPutFollowAction.java

```diff
@@ -128,93 +131,123 @@ private void createFollowerIndex(
-                    if (result) {
-                        initiateFollowing(request, listener);
-                    } else {
-                        listener.onResponse(new PutFollowAction.Response(true, false, false));
-                    }
-                },
-                listener::onFailure);
-        // Can't use create index api here, because then index templates can alter the mappings / settings.
-        // And index templates could introduce settings / mappings that are incompatible with the leader index.
-        clusterService.submitStateUpdateTask("create_following_index", new AckedClusterStateUpdateTask<Boolean>(request, handler) {
+        String remoteCluster = request.getRemoteCluster();
 
+        ActionListener<RestoreSnapshotResponse> restoreCompleteHandler = new ActionListener<RestoreSnapshotResponse>() {
             @Override
-            protected Boolean newResponse(final boolean acknowledged) {
-                return acknowledged;
+            public void onResponse(RestoreSnapshotResponse restoreSnapshotResponse) {
+                RestoreInfo restoreInfo = restoreSnapshotResponse.getRestoreInfo();
+                if (restoreInfo.failedShards() == 0) {
+                    initiateFollowing(request, listener);
+                } else {
+                    listener.onFailure(new ElasticsearchException("failed to restore [" + restoreInfo.failedShards() + "] shards"));
+                }
             }
 
             @Override
-            public ClusterState execute(final ClusterState currentState) throws Exception {
-                String followIndex = request.getFollowRequest().getFollowerIndex();
-                IndexMetaData currentIndex = currentState.metaData().index(followIndex);
-                if (currentIndex != null) {
-                    throw new ResourceAlreadyExistsException(currentIndex.getIndex());
+            public void onFailure(Exception e) {
+                listener.onFailure(e);
+
+            }
+        };
+
+        client.admin().cluster().preparePutRepository(remoteCluster).setType(RemoteClusterRepository.TYPE)
+            .execute(new ActionListener<AcknowledgedResponse>() {
+                @Override
+                public void onResponse(AcknowledgedResponse acknowledgedResponse) {
+                    if (acknowledgedResponse.isAcknowledged()) {
+                        Settings.Builder settingsBuilder = Settings.builder()
+                            .put(CcrSettings.CCR_FOLLOWING_INDEX_SETTING.getKey(), true)
+                            // TODO: Figure out what to do with private setting SETTING_INDEX_PROVIDED_NAME
+                            .put(IndexMetaData.SETTING_INDEX_PROVIDED_NAME, request.getFollowRequest().getFollowerIndex())
+                            // Overwriting UUID here, because otherwise we can't follow indices in the same cluster
+                            .put(IndexMetaData.SETTING_INDEX_UUID, UUIDs.randomBase64UUID())
```

Comment:
Is this necessary? AFAICS RestoreService is already generating a fresh UUID (the only exception being if there is an existing closed index).